src/category_theory/quotient.lean

```diff
@@ -0,0 +1,101 @@
+import category_theory.functor
+import category_theory.eq_to_hom
+
+/-
+  This file defines quotient categories. Given a category and an arbitrary family of relations
+  on its homsets, thought of as identifying some homs, we construct a new category on an
+  identical type of objects, where related homs are identified.
+
+    Note that we do *not* assume that our relation behaves well under composition, or that it is
+  an equivalence relation.
+-/
+
+universes v v₁ u u₁
+
+namespace category_theory
+
+variables
+  {C : Type u} [category.{v} C]
+  (r : Π (a b : C), (a ⟶ b) → (a ⟶ b) → Prop)
+include r
+
+/-- The type of objects in the quotient category --/
+structure quotient := (to_C : C)
+def of_C (a : C) : quotient r := ⟨r, a⟩
+
+namespace quotient
+
+/- Generates the closure of a family of relations wrt composition from left and right
+  If m₁ and m₂ are identified, then fm₁g and fm₂g should also be identified. -/
+inductive ccl {s t : C} : (s ⟶ t) → (s ⟶ t) → Prop
+| intro {a b} (f : s ⟶ a) (m₁ m₂ : a ⟶ b) (g : b ⟶ t) (h : r _ _ m₁ m₂) :
+  ccl (f ≫ m₁ ≫ g) (f ≫ m₂ ≫ g)
+
+lemma comp_left {a b c : C} (f : a ⟶ b) : Π (g₁ g₂ : b ⟶ c) (h : ccl r g₁ g₂),
+  ccl r (f ≫ g₁) (f ≫ g₂)
+| _ _ ⟨x, m₁, m₂, y, h⟩ := by simpa using ccl.intro (f ≫ x) m₁ m₂ y h
+
+lemma comp_right {a b c : C} (g : b ⟶ c) : Π (f₁ f₂ : a ⟶ b) (h : ccl r f₁ f₂),
+  ccl r (f₁ ≫ g) (f₂ ≫ g)
+| _ _ ⟨x, m₁, m₂, y, h⟩ := by simpa using ccl.intro x m₁ m₂ (y ≫ g) h
+
+def hom (s t : quotient r) := quot $ @ccl C _ r s.to_C t.to_C
+
+-- We get well-defined composition on the quotient because of comp_left and comp_right
+def comp {a b c : quotient r} : hom r a b → hom r b c → hom r a c :=
+λ hf hg, quot.lift_on hf ( λ f, quot.lift_on hg (λ g, quot.mk _ (f ≫ g))
+  (λ g₁ g₂ h, quot.sound $ comp_left r f g₁ g₂ h) )
+  (λ f₁ f₂ h, quot.induction_on hg $ λ g, quot.sound $ comp_right r g f₁ f₂ h)
+
+@[simp]
+lemma hcomp_mk {a b c : quotient r} (f : a.to_C ⟶ b.to_C) (g : b.to_C ⟶ c.to_C) :
```

Comment:
What does the name signify? Maybe this should be `comp_mk_mk`?

Reply:
The h currently means nothing. How about `comp_mk`? I stole this from `free_group.mul_mk`.